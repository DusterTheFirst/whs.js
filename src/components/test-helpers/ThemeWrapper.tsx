/*!
 * Copyright (C) 2018-2020  Zachary Kohnen (DusterTheFirst)
 */

import React from "react";
import { ThemeProvider } from "styled-components/native";
import { lightTheme } from "../../styles/theme";

/** The wrapper around a tested component to provide it with a theme */
const ThemeWrapper: React.FC = ({ children }) => (
    <ThemeProvider theme={lightTheme}>{children}</ThemeProvider>
);

export default ThemeWrapper;