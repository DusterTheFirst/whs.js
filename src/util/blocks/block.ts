/*!
 * Copyright (C) 2018-2019  Zachary Kohnen (DusterTheFirst)
 */

export enum Block {
    A = "a",
    B = "b",
    C = "c",
    D = "d",
    E = "e",
    F = "f"
}