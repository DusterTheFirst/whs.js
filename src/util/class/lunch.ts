/*!
 * Copyright (C) 2018-2019  Zachary Kohnen (DusterTheFirst)
 */

import { SchoolDay } from "../calendar/types";

/** The lunches for a day */
export enum Lunch {
    First,
    Second,
    Third
}

/** The lunches mapped by the school day. */
export type DayLunchMap = {
    [D in SchoolDay]: Lunch;
};