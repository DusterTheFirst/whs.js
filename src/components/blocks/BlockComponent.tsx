/*!
 * Copyright (C) 2018-2020  Zachary Kohnen (DusterTheFirst)
 */

import { BlockColor } from "@whsha/classes/v2/block";
import { IClass } from "@whsha/classes/v2/class";
import React, { memo } from "react";
import ClassComponent from "./ClassComponent";
import FreeComponent from "./FreeComponent";
import { ITimes } from "./times";

/** The parameters to pass to a BlockComponent */
interface IDisplayBlock extends ITimes {
    /** The class to display */
    clazz?: IClass;
    /** The block color */
    block: BlockColor;
}

/** A component that will display a block of many variety */
function BlockComponent({ end, start, clazz, block }: IDisplayBlock) {
    if (clazz === undefined) {
        return <FreeComponent start={start} end={end} block={block} />;
    } else {
        return (
            <ClassComponent
                start={start}
                end={end}
                clazz={clazz}
            />
        );
    }
}

export default memo(BlockComponent);