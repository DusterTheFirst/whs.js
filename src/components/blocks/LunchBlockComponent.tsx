/*!
 * Copyright (C) 2018-2020  Zachary Kohnen (DusterTheFirst)
 */

import { BlockColor } from "@whsha/classes/v2/block";
import { IClass } from "@whsha/classes/v2/class";
import { Lunch } from "@whsha/classes/v2/lunch";
import dayjs from "dayjs";
import { default as useCustomFormat } from "dayjs/plugin/customParseFormat";
import React, { memo } from "react";
import BlockComponent from "./BlockComponent";
import LunchComponent from "./LunchComponent";

dayjs.extend(useCustomFormat);

/** The parameters to pass to a LunchBlockComponent */
interface ILunchBlock {
    /** The class to display */
    lunch: Lunch;
    /** The class to display */
    clazz?: IClass;
    /** The block color */
    block: BlockColor;
    /** The timings for the classes */
    // times: {
    //     /** The first part of the lunch block */
    //     A: ITimes;
    //     /** The second part of the lunch block */
    //     B: ITimes;
    //     /** The third part of the lunch block */
    //     C: ITimes;
    // };
}

// FIXME: Correct timings
/** The timings for the lunch sections */
const LUNCH_TIMES = {
    A: {
        end: dayjs("11:22 AM", "h:mm A"),
        start: dayjs("10:55 AM", "h:mm A")
    },
    B: {
        end: dayjs("11:52 AM", "h:mm A"),
        start: dayjs("11:25 AM", "h:mm A")
    },
    C: {
        end: dayjs("12:22 PM", "h:mm A"),
        start: dayjs("11:55 AM", "h:mm A")
    }
};

/** A component that will display a block of many variety */
function LunchBlockComponent({ lunch, clazz, block }: ILunchBlock) {
    if (lunch === Lunch.First) {
        return (
            <>
                <LunchComponent start={LUNCH_TIMES.A.start} end={LUNCH_TIMES.A.end} />
                <BlockComponent start={LUNCH_TIMES.B.start} end={LUNCH_TIMES.C.end} block={block} clazz={clazz} />
            </>
        );
    } else if (lunch === Lunch.Second) {
        return (
            <>
                <BlockComponent start={LUNCH_TIMES.A.start} end={LUNCH_TIMES.A.end} block={block} clazz={clazz} />
                <LunchComponent start={LUNCH_TIMES.B.start} end={LUNCH_TIMES.B.end} />
                <BlockComponent start={LUNCH_TIMES.C.start} end={LUNCH_TIMES.C.end} block={block} clazz={clazz} />
            </>
        );
    } else if (lunch === Lunch.Third) {
        return (
            <>
                <BlockComponent start={LUNCH_TIMES.A.start} end={LUNCH_TIMES.B.end} block={block} clazz={clazz} />
                <LunchComponent start={LUNCH_TIMES.C.start} end={LUNCH_TIMES.C.end} />
            </>
        );
    } else {
        return (
            <BlockComponent start={LUNCH_TIMES.A.start} end={LUNCH_TIMES.C.end} block={block} clazz={clazz} />
        );
    }
}

export default memo(LunchBlockComponent);