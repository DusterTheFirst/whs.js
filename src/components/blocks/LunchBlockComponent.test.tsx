/*!
 * Copyright (C) 2018-2020  Zachary Kohnen (DusterTheFirst)
 */

import { BlockColor } from "@whsha/classes/v2/block";
import { Lunch } from "@whsha/classes/v2/lunch";
import dayjs from "dayjs";
import { default as useCustomFormat } from "dayjs/plugin/customParseFormat";
import React from "react";
import { render } from "react-native-testing-library";
import ThemeWrapper from "../test-helpers/ThemeWrapper";
import { dummyClass } from "./BlockComponent.test";
import LunchBlockComponent from "./LunchBlockComponent";

dayjs.extend(useCustomFormat);

describe("Tests different variations of classes", () => {
    it("Renders LunchBlockComponent with a class", () => {
        const comp = render(
            (
                <LunchBlockComponent
                    clazz={dummyClass}
                    block={BlockColor.Red}
                    lunch={Lunch.First}
                />
            ),
            { wrapper: ThemeWrapper }
        );

        expect(comp.toJSON()).toMatchSnapshot();
    });
    it("Renders LunchBlockComponent with a free", () => {
        const comp = render(
            (
                <LunchBlockComponent
                    block={BlockColor.Purple}
                    lunch={Lunch.Second}
                />
            ),
            { wrapper: ThemeWrapper });

        expect(comp.toJSON()).toMatchSnapshot();
    });
});
