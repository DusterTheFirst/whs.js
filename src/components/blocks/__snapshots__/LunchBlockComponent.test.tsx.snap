// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders BlockComponent with a class 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#efeff4",
        "justifyContent": "center",
        "paddingBottom": 10,
        "paddingLeft": 10,
        "paddingRight": 10,
        "paddingTop": 10,
        "width": "100%",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "flexWrap": "nowrap",
          "justifyContent": "space-between",
        },
      ]
    }
  >
    <Text
      classColor="Red"
      ellipsizeMode="middle"
      numberOfLines={1}
      style={
        Array [
          Object {
            "color": "#c93a3a",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "fontSize": 25,
            "fontWeight": "bold",
            "textAlign": "left",
          },
        ]
      }
    >
      No Name
    </Text>
    <View
      doFlex={false}
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "flex-end",
          },
        ]
      }
    >
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#808080",
              "paddingLeft": 5,
              "textAlign": "right",
            },
          ]
        }
      >
        7:30
         - 
        8:30 AM
      </Text>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "flexWrap": "nowrap",
          "justifyContent": "space-between",
          "marginTop": 5,
        },
      ]
    }
  >
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "color": "#808080",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "textAlign": "left",
          },
        ]
      }
    >
      No Teacher
    </Text>
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "color": "#808080",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "textAlign": "right",
          },
        ]
      }
    >
      No Room
    </Text>
  </View>
</View>
`;

exports[`Renders BlockComponent with a free 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#efeff4",
        "justifyContent": "center",
        "paddingBottom": 10,
        "paddingLeft": 10,
        "paddingRight": 10,
        "paddingTop": 10,
        "width": "100%",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "flexWrap": "nowrap",
          "justifyContent": "space-between",
        },
      ]
    }
  >
    <Text
      classColor="No Color"
      ellipsizeMode="middle"
      numberOfLines={1}
      style={
        Array [
          Object {
            "color": "#999999",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "fontSize": 25,
            "fontWeight": "bold",
            "textAlign": "left",
          },
        ]
      }
    >
      Free
    </Text>
    <View
      doFlex={false}
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "flex-end",
          },
        ]
      }
    >
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#808080",
              "paddingLeft": 5,
              "textAlign": "right",
            },
          ]
        }
      >
        7:30
         - 
        8:30 AM
      </Text>
    </View>
  </View>
</View>
`;

exports[`Tests different variations of classes Renders LunchBlockComponent with a class 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#efeff4",
          "justifyContent": "center",
          "paddingBottom": 10,
          "paddingLeft": 10,
          "paddingRight": 10,
          "paddingTop": 10,
          "width": "100%",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
            "flexWrap": "nowrap",
            "justifyContent": "space-between",
          },
        ]
      }
    >
      <Text
        ellipsizeMode="middle"
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#444444",
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "fontSize": 25,
              "fontWeight": "bold",
              "textAlign": "left",
            },
          ]
        }
      >
        Lunch
      </Text>
      <View
        doFlex={false}
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "flex-end",
            },
          ]
        }
      >
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#808080",
                "paddingLeft": 5,
                "textAlign": "right",
              },
            ]
          }
        >
          10:55
           - 
          11:22 AM
        </Text>
      </View>
    </View>
  </View>,
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#efeff4",
          "justifyContent": "center",
          "paddingBottom": 10,
          "paddingLeft": 10,
          "paddingRight": 10,
          "paddingTop": 10,
          "width": "100%",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
            "flexWrap": "nowrap",
            "justifyContent": "space-between",
          },
        ]
      }
    >
      <Text
        classColor="Red"
        ellipsizeMode="middle"
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#c93a3a",
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "fontSize": 25,
              "fontWeight": "bold",
              "textAlign": "left",
            },
          ]
        }
      >
        No Name
      </Text>
      <View
        doFlex={false}
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "flex-end",
            },
          ]
        }
      >
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#808080",
                "paddingLeft": 5,
                "textAlign": "right",
              },
            ]
          }
        >
          11:25
           - 
          12:22 PM
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
            "flexWrap": "nowrap",
            "justifyContent": "space-between",
            "marginTop": 5,
          },
        ]
      }
    >
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#808080",
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "textAlign": "left",
            },
          ]
        }
      >
        No Teacher
      </Text>
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#808080",
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "textAlign": "right",
            },
          ]
        }
      >
        No Room
      </Text>
    </View>
  </View>,
]
`;

exports[`Tests different variations of classes Renders LunchBlockComponent with a free 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#efeff4",
          "justifyContent": "center",
          "paddingBottom": 10,
          "paddingLeft": 10,
          "paddingRight": 10,
          "paddingTop": 10,
          "width": "100%",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
            "flexWrap": "nowrap",
            "justifyContent": "space-between",
          },
        ]
      }
    >
      <Text
        classColor="Purple"
        ellipsizeMode="middle"
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#8848ad",
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "fontSize": 25,
              "fontWeight": "bold",
              "textAlign": "left",
            },
          ]
        }
      >
        Free
      </Text>
      <View
        doFlex={true}
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexBasis": 0,
              "flexDirection": "row",
              "flexGrow": 1,
              "flexShrink": 1,
              "justifyContent": "flex-end",
            },
          ]
        }
      >
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#808080",
                "paddingLeft": 5,
                "textAlign": "right",
              },
            ]
          }
        >
          10:55
           - 
          11:22 AM
        </Text>
      </View>
    </View>
  </View>,
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#efeff4",
          "justifyContent": "center",
          "paddingBottom": 10,
          "paddingLeft": 10,
          "paddingRight": 10,
          "paddingTop": 10,
          "width": "100%",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
            "flexWrap": "nowrap",
            "justifyContent": "space-between",
          },
        ]
      }
    >
      <Text
        ellipsizeMode="middle"
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#444444",
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "fontSize": 25,
              "fontWeight": "bold",
              "textAlign": "left",
            },
          ]
        }
      >
        Lunch
      </Text>
      <View
        doFlex={false}
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "flex-end",
            },
          ]
        }
      >
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#808080",
                "paddingLeft": 5,
                "textAlign": "right",
              },
            ]
          }
        >
          11:25
           - 
          11:52 AM
        </Text>
      </View>
    </View>
  </View>,
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#efeff4",
          "justifyContent": "center",
          "paddingBottom": 10,
          "paddingLeft": 10,
          "paddingRight": 10,
          "paddingTop": 10,
          "width": "100%",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
            "flexWrap": "nowrap",
            "justifyContent": "space-between",
          },
        ]
      }
    >
      <Text
        classColor="Purple"
        ellipsizeMode="middle"
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#8848ad",
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "fontSize": 25,
              "fontWeight": "bold",
              "textAlign": "left",
            },
          ]
        }
      >
        Free
      </Text>
      <View
        doFlex={true}
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexBasis": 0,
              "flexDirection": "row",
              "flexGrow": 1,
              "flexShrink": 1,
              "justifyContent": "flex-end",
            },
          ]
        }
      >
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#808080",
                "paddingLeft": 5,
                "textAlign": "right",
              },
            ]
          }
        >
          11:55
           - 
          12:22 PM
        </Text>
      </View>
    </View>
  </View>,
]
`;
